File: src/main/java/katas/java/bsearch/BSearch1_.java
package katas.java.bsearch;

import org.junit.Test;

import java.util.Arrays;
import java.util.List;

import static java.util.Arrays.asList;
import static org.hamcrest.CoreMatchers.equalTo;
import static org.junit.Assert.assertThat;

/**
 * User: DKandalov
 */
public class BSearch1_ {
    @Test
    public void shouldFindElementPositionInAList() {
        assertThat(find(1, Arrays.<Integer>asList()), equalTo(-1));

        assertThat(find(0, asList(1)), equalTo(-1));
        assertThat(find(1, asList(1)), equalTo(0));
        assertThat(find(2, asList(1)), equalTo(-1));

        assertThat(find(0, asList(1, 2)), equalTo(-1));
        assertThat(find(1, asList(1, 2)), equalTo(0));
        assertThat(find(2, asList(1, 2)), equalTo(1));
        assertThat(find(3, asList(1, 2)), equalTo(-1));

        assertThat(find(0, asList(1, 2, 3)), equalTo(-1));
        assertThat(find(1, asList(1, 2, 3)), equalTo(0));
        assertThat(find(2, asList(1, 2, 3)), equalTo(1));
        assertThat(find(3, asList(1, 2, 3)), equalTo(2));
        assertThat(find(4, asList(1, 2, 3)), equalTo(-1));
    }

    private int find(int value, List<Integer> list) {
        int from = 0;
        int to = list.size();

        while (to - from != 0) {
            int midPos = (to + from) / 2;
            int midValue = list.get(midPos);

            if (midValue == value) {
                return midPos;
            } else if (value < midValue) {
                to = midPos;
            } else { // value > midValue
                from = midPos + 1;
            }
        }
        return -1;
    }
}

File: kotlin/src/katas/kotlin/binarysearch/BSearch1.kt
package katas.kotlin.binarysearch

import com.natpryce.hamkrest.assertion.assertThat
import com.natpryce.hamkrest.equalTo
import org.junit.Test

class BSearch1 {

    @Test fun `find index of item in list`() {
        listOf<Int>().let {
            assertThat(it.binarySearch(1), equalTo(-1))
        }

        listOf(1).let {
            assertThat(it.binarySearch(0), equalTo(-1))
            assertThat(it.binarySearch(1), equalTo(0))
            assertThat(it.binarySearch(2), equalTo(-1))
        }

        listOf(1, 2, 3).let {
            assertThat(it.binarySearch(0), equalTo(-1))
            assertThat(it.binarySearch(1), equalTo(0))
            assertThat(it.binarySearch(2), equalTo(1))
            assertThat(it.binarySearch(3), equalTo(2))
            assertThat(it.binarySearch(4), equalTo(-1))
        }
    }

    private fun <T : Comparable<T>> List<T>.binarySearch(value: T): Int {
        var from = 0
        var to = size
        while (from < to) {
            val midIndex = (from + to) / 2
            if (this[midIndex] == value) {
                return midIndex
            } else if (value < this[midIndex]) {
                to = midIndex
            } else {
                from = midIndex + 1
            }
        }
        return -1
    }
}
